### ISSUE TITLE
Add an option to customize conftest.py

### ISSUE THREAD

**Issue by davidedelvento** [pytest-dev/pytest#1569]
It would be very nice if I could do something like:

`py.test --conf=blah.py`

and have `py.test` run like if `blah.py` was named `conftest.py` and the latter (if existed) would be ignored. Of course invoking just:

`py.test`

would make everything run like it does now.

Looking at the source code, this would require change in just a handful of places [1] and would make switching between different conftests much more convenient than it is now

I'll be happy to do the change myself and submit a PR if you guys think this is a generally desirable feature but no one else wants to dedicate the time to implement it.

[1] namely:
* _pytest/config.py
* _pytest/doctest.py
* _pytest/helpconfig.py
* _pytest/main.py
* _pytest/python.py
* and of course in the main argument parsing code

**Comment by nicoddemus** [comment]
Could you explain why you need this change in the first place? I'm curious, because it seems to me to use a non-standard name for conftest files would only confuse other developers.

Having said that, instead of a command line option I think should be a `pytest.ini` option, because if a test suite chooses to change the standard name of `conftest` files then forgetting to pass that option would always break the test suite.

**Comment by nicoddemus** [comment]
(TBH I'm not certain having this option in the first place is a good idea)

**Comment by davidedelvento** [comment]
Main reason why I need it, and I need *not* in a ini file (so I would not contribute to that), is that I am using py.test to run a test suite that I am writing to test non-python things. Something similar to [what is officially documented](https://pytest.org/latest/example/nonpython.html). However, my code is complex enough, that I am also writing it in TDD, with python tests against it, which I run with py.test as well. So in the same directory structure I would like to run:

`py.test  # for the normal python tests of the test framework itself` 

and/or (possibly concurrently in another shell)

`py.test --conf=yaml_conftest.py # to run the equivalent of the example linked above` 

So far I have a python script which does:

    from shutil import copyfile
    try:
        copyfile("shark_pytest_conf.py", "conftest.py")
        pytest.main(str(mydir) + string_options)
    finally:
        os.remove("conftest.py")

So intead of the `py.test --conf` I run the script and that's good enough for my purpose (I can live without the concurrency), but I thought that having the option to select the conftest.py file on command line instead of messing up with files would be better, and useful to at least a few other folks (there are a few requests about it on SO)

**Comment by RonnyPfannschmidt** [comment]
whats preventing you from having a conftest at the root that just adds a option and picks plugins based on the option

a conftest is just the convention for a plugin that is specific for a folder, and your use-case screams pick custom plugins and register them

**Comment by RonnyPfannschmidt** [comment]
also please describe the problem you want to solve, i'm about 90% certain there already is a builtin solution and i dont want to suffer the XY Communication Problem

**Comment by davidedelvento** [comment]
@RonnyPfannschmidt 

- I described the problem I want to solve above, what part of that is not clear?
- What prevents me form having a conftest that adds an option to either ignore itself or not? Ignorance on how to do it! Only now that you mentioned it, I tried to run a `pytest_collect_file` which returns nothing and (to my great surprise) realized that it cause to py.test running its normal tests as opposed to running nothing at all. Is that documented, official behavior, or just side effect? If it's how it's supposed to work, I guess I can put it together with [adding an option](https://pytest.org/latest/example/parametrize.html) and achieve what I want.

That said, I agree that not every single "feature" is worth adding in a piece of software and that's why I opened this issue to discuss what you guys think about this before starting to work on it and send the PR.

Thanks for the feedback

**Comment by nicoddemus** [comment]
Thanks for the clarification @davidedelvento 

I think for your use case @RonnyPfannschmidt's idea of activating a plugin in the command line which then modifies pytest's normal behavior is the way to go. For an example of plugin which does something similar, [pytest-flakes](https://github.com/fschulze/pytest-flakes) comes to mind.

Sorry for not elaborating further, I'm short on time right now.

**Comment by RonnyPfannschmidt** [comment]
@davidedelvento your "use-case" contains a concrete implementation idea, which means its NOT a use case, its a implementation idea

a use-case describes the wanted behavior from the outside without implementation details being present
simply pretend you had no idea what py,test can do and "just" (which canbe surprisingly hard, i have been there often as well) describe the effects you want to see

**Comment by davidedelvento** [comment]
@RonnyPfannschmidt I'm sorry I thought my comment https://github.com/pytest-dev/pytest/issues/1569#issuecomment-221626784 in response to @nicoddemus clarified the problem I want to solve. Let me rephrase it as abstract as I can get, without any implementation details (other than "run py.test", ok? :-) 

I need to run py.test in two "modes":

1. Normal mode with autodiscovery for python tests (because I want to run the tests for the code I'm writing to do 2. below)

2. A "special" mode to run non-python tests, which at the moment I've implemented similarly to that Yaml example, but I'd be happy to refactor in other ways (how??) if it's worth the effort. 

A strong preference is that I do not want to deal with files for various reasons (e.g. moving conftest.py out of the way before running 1.) even though that is what I am doing right now as I described in a previous comment.

Good enough as use case? 

If so, let me go back to implementation details. Actually I do like a lot your idea of adding a conftest setting to create a command line option to switch between behavior 1 and 2 when invoking py.test -- The only thing that stops me to going that route right now is that I am not sure how best to "force" the behavior 1 when appropriate. I tried to have a `pytest_collect_file` into my conftest, that returns nothing at all, and that makes py.test running its normal tests as my requirement 1. (which is very good). But the more natural behavior to me seems that in those circumstances it should run nothing at all. Do you know if this behavior is an accident, or a deliberate design choice? I don't want to be stuck with something that may dramatically change in a future version of py.test (e.g. if a developer decides that `pytest_collect_file` returning nothing should force py.test do nothing):  that would preventing me from upgrading and I'll probably need/want to. Maybe a more robust choice would be that my `pytest_collect_file` (when the appropriate command line switch is activated) instead of returning nothing would call py.test version of that function and return whatever it gets from it. But I could not find anything explaining that in the documentation. 

Any other ideas on how to implement my use case?

Thanks!

**Comment by RonnyPfannschmidt** [comment]
if i recall correctly (untested) the most simple way is

write the code for 2 in a python module,

then in the conftest, use `pytest_plugins= [ 'the.module']`

and follow up by that you can use `py.test -p no:the.module` to disable the plugin

**Comment by davidedelvento** [comment]
This worked like a charm, thanks so much for the suggestion.

**Comment by gtristan** [comment]
I spent the greater part of a year wondering what this `conftest.py` file was doing in the root of my project (which landed there for the purpose of adding a session-wide fixture). I would very much appreciate a way to explicitly specify the filename in my `pytest` invocations or configuration, and at least be able to store it in the subdirectory containing my tests.

My first intuition, going by the name of the file, was that the file must be a part of the installation process, used to test host configurations and decide what directories things should be installed to or what support to enable on the target host - and I wondered why I have to maintain this file on top of all the rest of the build/install machinery.

A next reflex while wondering what is this files purpose is of course to `grep` my codebase, but then it looks like I can safely remove it; *nothing* is referring to this `conftest.py` file so it appears to be deadcode.

After removing it, tests break - now I am wondering how did this stuff get imported into a test ? Is it some plugin that is reading this file ?

The last thing on my mind was to expect for `pytest` proper to do black magic behind the scenes, and behave differently because of the existence of a file I did not even explicitly hand to `pytest`, it was very frustrating to reach all the way to google and have the surprise that this is a magical `pytest` detail.

At the very least, it would be nice to name it something sensible, like `pytest_globals.py` or *something* with `pytest` in the name.

**Comment by nicoddemus** [comment]
Hi @gtristan,

Sorry to hear you had so much trouble with your `conftest.py` file. While I understand that it is somewhat implicit (I remember that I did not like the name `conftest` back then when I started to look into pytest), changing it now would be impossible without breaking "the world". 

As for this being configurable, it actually defeats the purpose IMO: if one already knows enough about pytest that they know about the configuration option, they probably already know about `conftest.py` files.

**Comment by gtristan** [comment]
> changing it now would be impossible without breaking "the world".

Yes that is obvious, it *could* have another well known alias (but it would be much preferable to be allowed to place this file in the `tests/` subdirectory and reduce clutter from the project root, which already has `tox.ini`, `setup.py`, `setup.cfg`, `versioneer.py`).

> As for this being configurable, it actually defeats the purpose IMO: if one already knows enough about pytest that they know about the configuration option, they probably already know about conftest.py files.

Of course, but I believe you are looking at the problem backwards; we should not assume that someone working on a given project has deep knowledge of `pytest` in particular, (we have had close to 20 full time employees working on [BuildStream](https://gitlab.com/BuildStream/buildstream/) at one point), not counting fly by contributors who want to efficiently submit patches to scratch their own itch.

A goal for a project maintainer is to always lower the bar of entry as much as possible, for various reasons:
* The maintainer today might not be the maintainer tomorrow
* Fly by contributors or short term contractors need to be able to easily submit patches with regression tests, we should not give them an environment which requires detective work to figure out

**Comment by RonnyPfannschmidt** [comment]
its already possible to place it in the tests folder (presuming you don't use doctests or  cli args as well)
changing the name at this point in time is highly unlikely and very expensive
similarly it does not look like the cost of making it configurable is worth the stream of additional confusion/issues

a -10 from me

**Comment by gtristan** [comment]
@RonnyPfannschmidt ok look at it this way: `pytest` is reading a file that I did not explicitly hand it, seeing as it does not start with `.pytest...` this is already a bit alarming. What assurance do I have that `pytest` will behave stably and predictably from one version to another ? What if there is another file that it one day decides to read, and I have one under the same name which is used for something different ?

> its already possible to place it in the tests folder (presuming you don't use doctests or cli args as well)

I did try to place it there, didnt work, but I didnt really spend a lot of time trying to figure out if it was intended to work or not, just placed it there and thought the sneaky secret `conftest.py` scanner might find it and pick it up.

Not sure what you mean by cli args, we certainly invoke `pytest` with cli args, and these will differ depending on whether you invoke `setup.py test` directly (to explicitly test against a host environment, useful for downstream distro packagers) or whether you invoke `tox` (more useful for developers and in CI).

> changing the name at this point in time is highly unlikely and very expensive

Not really asking for this, as I said, an alias would probably be the cheapest thing, but still not really great.

> similarly it does not look like the cost of making it configurable is worth the stream of additional confusion/issues

I can see that it opens the door to some different constructs (one might invoke `pytest` in a loop while specifying different `conftest.py` files, for example, that might change how people use `pytest` and bring different kinds of issues) - the benefit is that you can setup your project such that everything is explicit, and nothing sneaky and unpredictable is happening behind your back.

**Comment by gtristan** [comment]
I will try harder to get my project to work with `conftest.py` in the `tests/` subdir; frankly speaking this will at least be enough of a hint that the file is related to test.

**Comment by RonnyPfannschmidt** [comment]
> @RonnyPfannschmidt ok look at it this way: `pytest` is reading a file that I did not explicitly hand it, seeing as it does not start with `.pytest...` this is already a bit alarming. What assurance do I have that `pytest` will behave stably and predictably from one version to another ? What if there is another file that it one day decides to read, and I have one under the same name which is used for something different ?
> 

its a well documented part of the system/plugin system, if there should ever be an expanse like that it will be well documented and in a feature release, for control over that there is version pinning.

> > its already possible to place it in the tests folder (presuming you don't use doctests or cli args as well)
> 
> I did try to place it there, didnt work, but I didnt really spend a lot of time trying to figure out if it was intended to work or not, just placed it there and thought the sneaky secret `conftest.py` scanner might find it and pick it up.
> 
> Not sure what you mean by cli args, we certainly invoke `pytest` with cli args, and these will differ depending on whether you invoke `setup.py test` directly (to explicitly test against a host environment, useful for downstream distro packagers) or whether you invoke `tox` (more useful for developers and in CI).
> 
the `pytest_addopts(parser)` hook which allows you to add own command line arguments for pytest

> > changing the name at this point in time is highly unlikely and very expensive
> 
> Not really asking for this, as I said, an alias would probably be the cheapest thing, but still not really great.
> 
> > similarly it does not look like the cost of making it configurable is worth the stream of additional confusion/issues
> 
> I can see that it opens the door to some different constructs (one might invoke `pytest` in a loop while specifying different `conftest.py` files, for example, that might change how people use `pytest` and bring different kinds of issues) - the benefit is that you can setup your project such that everything is explicit, and nothing sneaky and unpredictable is happening behind your back.

i do not share your perception there, there is no need for more than one config files to implement that use-case and there already are multiple ways to do it cleanly (and `conftest.py` is one of them)

**Comment by philippefutureboy** [comment]
@RonnyPfannschmidt I come from JS land, where we have the excellent jest, and jest does not load files "by convention".
One has to explicitly configure it via flags or via a configuration section in the local `package.json` or via a `--config` flag.
Jest has a `setupFiles` flag (https://jestjs.io/docs/en/configuration#setupfiles-array) which specifies which config files to load.

While I understand your position as a maintainer of a very widely used package and your responsibility of not breaking the world for those using your software as part of their daily operations, I think there is place for a non-breaking `--config` flag or `--setup_files` flag à la jest that allows to explicitly load config files. Mixed with `--noconftest`, this could have the effect desired by a lot of the commenters above.

Have a great day!

**Comment by RonnyPfannschmidt** [comment]
> @RonnyPfannschmidt I come from JS land, where we have the excellent jest, and jest does not load files "by convention".
> One has to explicitly configure it via flags or via a configuration section in the local `package.json` or via a `--config` flag.
> Jest has a `--setupFiles` flag (https://jestjs.io/docs/en/configuration#setupfiles-array) which specifies which config files to load.
> 
> While I understand your position as a maintainer of a very widely used package and your responsibility of not breaking the world for those using your software as part of their daily operations, I think there is place for a non-breaking `--config` flag or `--setup_files` flag à la jest that allows to explicitly load config files. Mixed with `--noconftest`, this could have the effect desired by a lot of the commenters above.
> 
> Have a great day!

For the actual config file pytest has support for pyproject.toml by now and supports passing custom files as well. Jo

this also seems unrelated to this issue that has been closed years ago?!?

**Comment by philippefutureboy** [comment]
Oh! Well it must have passed under my nose. I couldn't find anything in the docs. Thanks for saying :)
